### ISSUE TITLE
Regression: crash deep in XmlSerializationReaderILGen when creating XmlSerializer (netcore 3.1)

### ISSUE THREAD

**Issue by uxmal** [dotnet/runtime#43675]
I'm trying to port my project from .NET Framework 4.7.2 to netcore3.1. I have over 10,000 NUnit 3 tests, of which all are passing on .NET Framework 4.7.2. The ported unit tests are passing except for a handful. The majority of these failures are caused by the issue I'm reporting here. I am running the unit tests from the NUnit3 Test runner adapter hosted inside of Visual Studio 2019, running in 64-bit Windows 10.  `dotnet version` reports `3.1.201`

I am creating an `XmlSerializer` using an instance of `XmlAttributeOverrides`:
```C#
new XmlSerializer(rootType, attrOverrides);
```
Deep inside the constructor of `XmlSerializer` there is a crash:
```
  Message: 
    System.Collections.Generic.KeyNotFoundException : The given key 'QualifierValues1' was not present in the dictionary.
  Stack Trace: 
    Dictionary`2.get_Item(TKey key)
    XmlSerializationReaderILGen.MakeUnique(EnumMapping mapping, String name)
    XmlSerializationReaderILGen.WriteHashtable(EnumMapping mapping, String typeName, MethodBuilder& get_TableName)
    XmlSerializationReaderILGen.WriteEnumMethod(EnumMapping mapping)
    XmlSerializationReaderILGen.GenerateMethod(TypeMapping mapping)
    XmlSerializationILGen.GenerateReferencedMethods()
    XmlSerializationReaderILGen.GenerateEnd(String[] methods, XmlMapping[] xmlMappings, Type[] types)
    TempAssembly.GenerateRefEmitAssembly(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace)
    TempAssembly.ctor(XmlMapping[] xmlMappings, Type[] types, String defaultNamespace, String location)
    XmlSerializer.GenerateTempAssembly(XmlMapping xmlMapping, Type type, String defaultNamespace, String location)
    XmlSerializer.ctor(Type type, XmlAttributeOverrides overrides, Type[] extraTypes, XmlRootAttribute root, String defaultNamespace, String location)
    XmlSerializer.ctor(Type type, XmlAttributeOverrides overrides)
```
This is a regression from .NET Framework 4.7.2, where such a crash is not happening.

The git branch on which this error manifests itself is here: https://github.com/uxmal/reko/tree/netStandard-ex

To reproduce:
* git clone https://github.com/uxmal/reko
* git checkout netStandard-ex
* cd src
* Open the project Reko-decompiler.sln with VS 2019
* Select the "Debug" configuration and the "x64" platform.
* Build the solution
* Run the unit test Reko.UnitTests.Core.Serialization.SerializedServiceTests.SserSerialize

Result: stack trace above

Expected: no crash

I can provide the binaries upon request; I'm hesitant to attach them here since they're several megabytes.

I'm unaware of any workarounds. Unfortunately, this is blocking my move off .NET Framework 4.7.2. I'm strongly motivated to assist in troubleshooting this issue.

**Comment by kevingosse** [comment]
```csharp
        private string? MakeUnique(EnumMapping mapping, string name)
        {
            string uniqueName = name;
            EnumMapping? m;
            if (Enums.TryGetValue(uniqueName, out m))
            {
                if (m == mapping)
                {
                    // we already have created the hashtable
                    return null;
                }
                int i = 0;
                while (m != null)
                {
                    i++;
                    uniqueName = name + i.ToString(CultureInfo.InvariantCulture);
                    m = Enums[uniqueName];
                }
            }
            Enums.Add(uniqueName, mapping);
            return uniqueName;
        }
```

The issue is the ` m = Enums[uniqueName];` line. Originally, `Enums` was a HashTable, which means that the statement would return null when no key is found. At some point it has been changed to a Dictionary (https://github.com/dotnet/runtime/commit/1411b5c524c1e35b5f503e97947ee4fa9b5021ee), which changes the behavior of the indexer. I believe this should be replaced with a call to `TryGetValue`.

**Comment by StephenMolloy** [comment]
This has been fixed by #49892 for .Net 6.0. It looks like the original report for this bug has been worked around. I would also take a moment to point out that if the original repro passed the namespace for the root type being serialized, then this issue could be avoided. It seems odd that the repro only applies a namespace to a small handful of classes in the object graph.

Since this no longer appears to be blocking adoption, I don't think this meets servicing bars for bringing this back. If anyone can make a case for bringing this fix back to 5.0 or 3.1, let us know and we can reconsider servicing this fix. Thanks.
